ray-curator/ray_curator/core/utils.py

```diff
@@ -0,0 +1,119 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import socket
+import subprocess
+from typing import TYPE_CHECKING
+
+import ray
+from loguru import logger
+
+from ray_curator.core.constants import (
+    DEFAULT_RAY_AUTOSCALER_METRIC_PORT,
+    DEFAULT_RAY_DASHBOARD_METRIC_PORT,
+)
+
+if TYPE_CHECKING:
+    import loguru
+
+
+def get_free_port(start_port: int, get_next_free_port: bool = True) -> int:
+    """Checks if start_port is free.
+    If not, it will get the next free port starting from start_port if get_next_free_port is True.
+    Else, it will raise an error if the free port is not equal to start_port.
+    """
+    for port in range(start_port, 65535):
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+            # SO_REUSEADDR to avoid TIME_WAIT issues on some OSes
+            s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+            try:
+                s.bind(("localhost", port))
+                # If bind succeeds, port is free
+                return port  # noqa: TRY300
+            except OSError:
+                if not get_next_free_port and port == start_port:
+                    msg = f"Port {start_port} is already in use. Please provide a different port."
+                    raise RuntimeError(msg)  # noqa: B904
+                continue
+    msg = f"No free port found between {start_port} and 65535"
+    raise RuntimeError(msg)
+
+
+def _logger_custom_serializer(
+    _: "loguru.Logger",
+) -> None:
+    return None
+
+
+def _logger_custom_deserializer(
+    _: None,
+) -> "loguru.Logger":
+    # Initialize a default logger
+    return logger
+
+
+def init_cluster(  # noqa: PLR0913
+    ray_port: int,
+    ray_temp_dir: str,
+    ray_dashboard_port: int,
+    ray_metrics_port: int,
+    ray_client_server_port: int,  # noqa: ARG001
+    ray_dashboard_host: str,
+    num_gpus: int | None = None,
+    num_cpus: int | None = None,
+    enable_object_spilling: bool = False,
+    block: bool = True,
```

Comment:
I dont think we should support non blocking mode today, lets remove this option for now please as if non blocking we can not kill it

Reply:
I think once ray allows to stop a sepcific server this might be used. I can add an assert that block is always true.

Reply:
Ok we are not passing `block` as a param from get_client class (which means it's not an option). But incase the user wants to use this util without `block`, I don't see a reason why they should not be able to ?